rpc/gnmi/gnmi-depth.md

```diff
@@ -0,0 +1,290 @@
+
+### 4.2 depth 1, path `/basket/fruits`
+
+When the path targets the list schema node, all elements of this list is returned with their children nodes
+
+```bash
+gnmic -e json_ietf get --path /basket/fruits --depth 1
+```
+
+```json
+[
+  {
+    "fruits": [
+      {
+        "colors": [
+          "red",
+          "yellow"
+        ],
+        "name": "apples",
+        "size": "XL"
+      },
+      {
+        "name": "orange",
+        "size": "M"
+      }
+    ]
+  }
+]
+```
+
+Again, please keep in mind that only leafs and leaf-lists are returned for every list element.
+
+### 4.3 depth 2, path `/basket`
+
+When the depth level is set to values >1, all elements from the path to the provided level value are returned in full with the last level including only leafs and leaf-lists.
+
+```bash
+gnmic -e json_ietf get --path /basket --depth 2
+```
+
+```json
+[
+  {
+    "broken": {
+      "reason": "too heavy"
+    },
+    "contents": [
+      "fruits",
+      "vegetables"
+    ],
+    "description": {
+      "fabric": "cotton"
+    },
+    "fruits": [
+      {
+        "colors": [
+          "red",
+          "yellow"
+        ],
+        "name": "apples",
+        "size": "XL"
+      },
+      {
+        "name": "orange",
+        "size": "M"
+      }
+    ]
+  }
+]
+```
+
+Here is what happens:
+
+<img width="640" alt="image" src="https://github.com/openconfig/gnmi/assets/5679861/331a20a5-ff19-4932-873a-7aff9b36dd62">
+
+The 1st level elements are returned, since depth level is 2.
+On the 2nd level we return only leafs and leaf-lists, hence the `.fruits.origin` is not present.
+
+## 5 Prior art
+
+Netconf standardized `max-depth` in [RFC 85226](https://datatracker.ietf.org/doc/html/rfc8526#section-3.1.1):
+
+> The "max-depth" parameter can be used by the client to limit the
+> number of subtree levels that are returned in the reply.
+
+The NETCONF way of using the max-depth differs in a sense that `depth=1` returns the element pointed by the path, but not its children. depth=2 returns children of the element pointed by the path.
+
+I find this behavior strange, as I don't see an operational reason to return the element itself when depth is 1.
+
+## 6 Summary
+
+We believe that the Depth extension has a generic applicability whilst not bein a burden for the implementation (henceforth no subtree filtering with XPath or anything of sorts).
+
+Yet it delivers important quality of life improvements for consuming systems that may get the required data nodes faster and with less processing time spent.
+
+This is assuming that cumulative time of fetching only leaf/leaf-lists values by the server is smaller than the recursive data retrieval combined with payload unmarshalling on the client side.
```

Comment:
Maybe rewrite or remove this section (the section "Purpose", it seems, already covers the use cases)